### PR TITLE
require extension option explicitly

### PR DESCRIPTION
**Pull request by erwinmombay**
get rid of thisOrThat pattern (tries to emulate function overloading) in favor of explicit option object requirement.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -63,128 +63,134 @@ var unminifiedRuntimeTarget = 'dist/amp.js';
...
-    options = hasCssOrOptions
-  }
-  extensions[name + '-' + version] = Object.assign({
+function declareExtension(name, version, options) {
+  const defaultOptions = {hasCss: false};
+  const extensionKey = `${name}-${version}`;
+  extensions[extensionKey] = Object.assign({
     name: name,
     version: version,
     hasCss: hasCss,
```
Comment (kristoferbaxter) [inline gulpfile.js]: Can remove this line, hasCss isn't available anymore in this scope!

File: gulpfile.js [hunk 2]
```diff
@@ -63,128 +63,134 @@ var unminifiedRuntimeTarget = 'dist/amp.js';
...
-    hasCss = hasCssOrOptions;
-  } else {
-    options = hasCssOrOptions
-  }
-  extensions[name + '-' + version] = Object.assign({
+function declareExtension(name, version, options) {
+  const defaultOptions = {hasCss: false};
+  const extensionKey = `${name}-${version}`;
+  extensions[extensionKey] = Object.assign({
     name: name,
```
Comment (kristoferbaxter) [inline gulpfile.js]: No requirement here, but can leverage Shorthand property names.

```
Object.assign({
  name,
  version,
}, defaultOptions, options);
```

**Comment by cramforce** [comment]
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/ampproject/amphtml/rev/pr-146607719cdc93309d1d6d18547312889f91caa5)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [lgtm.com](https://lgtm.com)*

File: gulpfile.js (hunk 1 above)
Comment (erwinmombay) [reply gulpfile.js]: thanks!

File: gulpfile.js (hunk 2 above)
Comment (erwinmombay) [reply gulpfile.js]: done.

File: gulpfile.js [hunk 3]
```diff
@@ -63,128 +63,134 @@ var unminifiedRuntimeTarget = 'dist/amp.js';
...
+
+/**
+ * @typedef {{
+ *   name: ?string,
+ *   version: ?string,
+ *   hasCss: ?boolean,
+ *   loadPriority: ?string
+ * }}
+ */
+const ExtensionOption;
```
Comment (kristoferbaxter) [inline gulpfile.js]: Looks like a leftover const definition. Mind removing?

**Comment by kristoferbaxter** [comment]
@rsimha-amp The extension I'm working on adds some additional options. While reviewing how to introduce – I noticed this pattern where the `declareExtension` function was inspecting the type the third argument was and then adapting to it.

@erwinmombay and I chatted about it and didn't see a reason why it should do this instead of passing a known typed argument. This PR is the result of that conversation.

**Comment by erwinmombay** [comment]
@rsimha-amp I'm just not a fan of the thisTypeOrAnotherType pattern (because javascript doesn't have method/function overloading). Unions are also OK but but this one tries to emulate method overloading instead which always turns ugly logically. (just look at the If else logic we had to remove)

File: gulpfile.js (hunk 3 above)
Comment (erwinmombay) [reply gulpfile.js]: this is required for jscompiler typechecking (it needs a reference to an identifier)

**Pull request merged**